tensorflow_recommenders_addons/dynamic_embedding/core/kernels/lookup_impl/lookup_table_op_gpu.h

```diff
@@ -98,6 +112,177 @@ class TableWrapper final : public TableWrapperBase<K, V> {
     table_->dump(d_key, d_val, offset, search_length, d_dump_counter, stream);
   }
 
+  Status save_to_file_system(OpKernelContext* ctx, const int64 value_dim,
+                             const string& filepath, const size_t buffer_size,
+                             cudaStream_t stream) override {
+    const size_t dim = static_cast<size_t>(value_dim);
+
+    FileSystem* fs;
+    const auto env = ctx->env();
+    TF_RETURN_IF_ERROR(env->GetFileSystemForFile(filepath, &fs));
+    std::unique_ptr<WritableFile> key_writer;
+    std::unique_ptr<WritableFile> value_writer;
+    const string key_filepath = filepath + ".keys";
+    const string value_filepath = filepath + ".values";
+    const string key_tmpfilepath = filepath + ".keys.tmp";
+    const string value_tmpfilepath = filepath + ".values.tmp";
+    TF_RETURN_IF_ERROR(fs->NewWritableFile(key_tmpfilepath, &key_writer));
+    TF_RETURN_IF_ERROR(fs->NewWritableFile(value_tmpfilepath, &value_writer));
+
+    size_t key_offset = 0;
+    size_t value_offset = 0;
+    const size_t value_len = sizeof(V) * dim;
+    const size_t key_buffer_byte_size = buffer_size * sizeof(K);
+    const size_t value_buffer_byte_size = buffer_size * value_len;
+    char* key_buffer = new char[key_buffer_byte_size];
+    char* value_buffer = new char[value_buffer_byte_size];
+
+    K* d_keys = nullptr;
+    V* d_values = nullptr;
+    size_t search_offset = 0;
+    size_t* d_dump_counter;
+    size_t dump_counter;
+    size_t table_capacity = get_capacity();
+
+    CUDA_CHECK(cudaMalloc(&d_keys, key_buffer_byte_size));
+    CUDA_CHECK(cudaMalloc(&d_values, value_buffer_byte_size));
+    CUDA_CHECK(cudaMalloc(&d_dump_counter, sizeof(size_t)));
+
+    size_t search_length = 0;
+    size_t total_saved = 0;
+    while (search_offset < table_capacity) {
+      if (search_offset + buffer_size >= table_capacity) {
+        search_length = table_capacity - search_offset;
+      } else {
+        search_length = buffer_size;
+      }
+      table_->dump(d_keys, (ValueType<V>*)d_values, search_offset,
+                   search_length, d_dump_counter, stream);
+      CUDA_CHECK(cudaStreamSynchronize(stream));
+
+      CUDA_CHECK(cudaMemcpyAsync(&dump_counter, d_dump_counter, sizeof(size_t),
+                                 cudaMemcpyDeviceToHost, stream));
+      CUDA_CHECK(cudaStreamSynchronize(stream));
+
+      if (dump_counter > 0) {
+        key_offset = dump_counter * sizeof(K);
+        value_offset = dump_counter * value_len;
+        CUDA_CHECK(cudaMemcpyAsync(key_buffer, d_keys, key_offset,
```

Comment:
The cudaMemcpyAsync dst, `key_buffer`, should be device accessible, you should allocate it by `cudaMallocManaged`